libs/langchain-google-common/src/chat_models.ts

```diff
@@ -60,14 +61,67 @@ class ChatConnection<AuthOptions> extends AbstractGoogleLLMConnection<
   BaseMessage[],
   AuthOptions
 > {
+
+  convertSystemMessageToHumanContent: boolean | undefined;
+
+  constructor(fields: GoogleAIBaseLLMInput<AuthOptions> | undefined, caller: AsyncCaller, client: GoogleAbstractedClient, streaming: boolean) {
+    super(fields, caller, client, streaming);
+    this.convertSystemMessageToHumanContent = fields?.convertSystemMessageToHumanContent;
+  }
+
+  get useSystemInstruction(): boolean {
+    return typeof this.convertSystemMessageToHumanContent === "boolean"
+      ? !this.convertSystemMessageToHumanContent
+      : this.computeUseSystemInstruction;
+  }
+
+  get computeUseSystemInstruction(): boolean {
+    // This works on models from April 2024 and later
+    //   Vertex AI: gemini-1.5-pro and gemini-1.0-002 and later
+    //   AI Studio: gemini-1.5-pro-latest
+    if (this.modelFamily === "palm") {
+      return false;
+    } else if (this.modelName === "gemini-1.0-pro-001") {
+      return false;
+    } else if (this.modelName.startsWith("gemini-pro-vision")) {
+      return false;
+    } else if (this.modelName.startsWith("gemini-1.0-pro-vision")) {
+      return false;
+    } else if (this.modelName === "gemini-pro" && this.platform === "gai") {
+      // on AI Studio gemini-pro is still pointing at gemini-1.0-pro-001
+      return false;
+    }
+    return true;
+  }
+
   formatContents(
     input: BaseMessage[],
     _parameters: GoogleAIModelParams
   ): GeminiContent[] {
     return input
-      .map((msg, i) => baseMessageToContent(msg, input[i - 1]))
-      .reduce((acc, cur) => [...acc, ...cur]);
+      .map((msg, i) =>
+        baseMessageToContent(msg, input[i - 1], this.useSystemInstruction)
+      )
+      .reduce((acc, cur) => {
+        // Filter out the system content, since those don't belong
+        // in the actual content.
+        const hasNoSystem = cur.every(content => content.role !== "system");
+        return hasNoSystem ? [...acc, ...cur] : acc;
+      }, []);
   }
+
+  formatSystemInstruction(
+    input: BaseMessage[],
+    _parameters: GoogleAIModelParams
+  ): GeminiContent {
+    const firstInput = input[0];
+    if (firstInput._getType() === "system" && this.useSystemInstruction) {
+      return baseMessageToContent(firstInput, undefined, true)[0];
```

Comment:
What if we had a system message in another part of the prompt? Google would throw an error?

Reply:
Google only accepts one system message (see the other notes about this).

In this code, it is just looking for it as the first message.
But I guess one could argue that any of these others are better approaches:
- Throws an error if one is found _not_ as the first message.
- Combine any of them found, in any position, into the system prompt

Thoughts?